File: truevfs-ext/truevfs-ext-pacemaker/src/main/scala/net/java/truevfs/ext/pacemaker/LruCache.scala
package net.java.truevfs.ext.pacemaker

import java.util.Map.Entry
import java.util.concurrent.locks.{Lock, ReadWriteLock, ReentrantReadWriteLock}
import java.util.{concurrent => juc}
import java.{util => ju}

import net.java.truecommons.shed.HashMaps

import scala.collection.JavaConverters._

/** A simple cache set with a least-recently-used (LRU) eviction strategy.
  * Note that unlike other caches, whenever an item gets evicted, it gets
  * added to a concurrent set which can get queried using the [[evicted]] method
  * for further processing, e.g. close resources.
  * This class is thread-safe.
  *
  * @tparam I the type of the items
  * @author Christian Schlichtherle
  */
private final class LruCache[I <: AnyRef](initialMaximumSize: Int) {

  private val _lock: ReadWriteLock = new ReentrantReadWriteLock
  private val _readLock = _lock.readLock
  private val _writeLock = _lock.writeLock

  private val _evicted: juc.ConcurrentMap[I, Boolean] = new juc.ConcurrentHashMap
  private val _cached = new CacheMap

  @volatile
  private var _maximumSize: Int = _

  maximumSize = initialMaximumSize

  def evicted: ju.Set[I] = _evicted.keySet

  def maximumSize: Int = _maximumSize
  def maximumSize_=(maximumSize: Int): Unit = {
    require(0 <= maximumSize)
    _maximumSize = maximumSize
  }

  def add(item: I): Boolean = writeLocked { _cached put (item, true) }

  /** Records access to the given mount point.
    * This method has no effect if the given mount point is not present in this
    * LRU cache.
    */
  def recordAccess(item: I): Unit = {
    // The lookup needs to be write-locked because the access-ordered cache map
    // may get structurally modified as a side effect.
    writeLocked { _cached get item }
  }

  def remove(item: I): Boolean = writeLocked { _cached remove item }

  def exists(predicate: I => Boolean): Boolean =
    readLocked { _cached exists predicate }

  private def readLocked[V]: (=> V) => V = locked[V](_readLock)

  private def writeLocked[V]: (=> V) => V = locked[V](_writeLock)

  private def locked[V](lock: Lock)(block: => V): V = {
    lock.lock()
    try {
      block
    } finally {
      lock.unlock()
    }
  }

  private final class CacheMap
    extends ju.LinkedHashMap[I, Boolean](HashMaps initialCapacity initialMaximumSize, 0.75f, true) {

    override def removeEldestEntry(entry: Entry[I, Boolean]): Boolean =
      if (size > maximumSize) {
        _evicted put (entry.getKey, entry.getValue)
        true
      } else {
        false
      }

    override def put(key: I, value: Boolean): Boolean = {
      _evicted remove key
      super.put(key, value)
    }

    override def remove(key: Object): Boolean = {
      _evicted remove key
      super.remove(key)
    }

    def exists(predicate: I => Boolean): Boolean = keySet.asScala exists predicate
  }
}

File: src/main/java/global/namespace/truevfs/ext/pacemaker/LruCache.java
/*
 * Copyright © 2005 - 2021 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package global.namespace.truevfs.ext.pacemaker;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import java.util.function.Predicate;
import java.util.function.Supplier;

import static global.namespace.truevfs.commons.shed.HashMaps.initialCapacity;

/**
 * A simple cache set with a least-recently-used (LRU) eviction strategy.
 * Note that unlike other caches whenever an item gets evicted, it gets added to a concurrent set which can get queried
 * using the {@link #evicted} method for further processing, e.g. to close resources.
 * <p>
 * This class is thread-safe.
 *
 * @param <T> the type of the items
 * @author Christian Schlichtherle
 */
final class LruCache<T> {

    private final ReadWriteLock lock = new ReentrantReadWriteLock();
    private final Lock readLock = lock.readLock();
    private final Lock writeLock = lock.writeLock();

    private final ConcurrentMap<T, Boolean> evicted = new ConcurrentHashMap<>();
    private final CacheMap cached = new CacheMap();

    private volatile int maximumSize;

    LruCache(final int initialMaximumSize) {
        this.maximumSize = initialMaximumSize;
    }

    Set<T> getEvictedView() { return evicted.keySet(); }

    int getMaximumSize() { return maximumSize; }

    void setMaximumSize(final int maximumSize) {
        if (0 > maximumSize) {
            throw new IllegalArgumentException();
        }
        this.maximumSize = maximumSize;
    }

    void add(T item) {
        writeLocked(() -> cached.put(item, true));
    }

    void remove(T item) {
        writeLocked(() -> cached.remove(item));
    }

    /**
     * Records access to the given mount point.
     * This method has no effect if the given mount point is not present in this LRU cache.
     */
    void recordAccess(T item) {
        // The lookup needs to be write-locked because the access-ordered cache map may get structurally modified as a
        // side effect:
        writeLocked(() -> cached.get(item));
    }

    boolean exists(Predicate<T> predicate) {
        return readLocked(() -> cached.exists(predicate));
    }

    private <U> U readLocked(Supplier<U> expr) {
        return locked(readLock, expr);
    }

    private <U> U writeLocked(Supplier<U> expr) {
        return locked(writeLock, expr);
    }

    private static <U> U locked(final Lock lock, final Supplier<U> expr) {
        lock.lock();
        try {
            return expr.get();
        } finally {
            lock.unlock();
        }
    }

    private final class CacheMap extends LinkedHashMap<T, Boolean> {

        private static final long serialVersionUID = 0;

        CacheMap() {
            super(initialCapacity(maximumSize), 0.75f, true);
        }

        @Override
        protected boolean removeEldestEntry(final Map.Entry<T, Boolean> eldest) {
            if (size() > maximumSize) {
                evicted.put(eldest.getKey(), eldest.getValue());
                return true;
            } else {
                return false;
            }
        }

        @Override
        public Boolean put(final T key, final Boolean value) {
            evicted.remove(key);
            return super.put(key, value);
        }

        @Override
        public Boolean remove(final Object key) {
            evicted.remove(key);
            return super.remove(key);
        }

        boolean exists(Predicate<T> predicate) {
            return keySet().stream().anyMatch(predicate);
        }
    }
}
